### PR TITLE
[MODORDERS-929] - Local shadow instance creation for linked instances

### PR DESCRIPTION
**Pull request by Abdulkhakimov**
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/service/consortium/ConsortiumConfigurationService.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.models.consortium.ConsortiumConfiguration;
+import org.folio.rest.core.RestClient;
+import org.folio.rest.core.models.RequestContext;
+import org.folio.rest.core.models.RequestEntry;
+
+import java.util.Optional;
+
+public class ConsortiumConfigurationService {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/service/consortium/ConsortiumConfigurationService.java:13]: consider having a cache to keep these values. They do no change often.

File: src/main/java/org/folio/helper/PurchaseOrderLineHelper.java [hunk 2]
```diff
@@ -884,6 +886,13 @@ private Future<Void> verifyDeleteAllowed(PoLine line, RequestContext requestCont
         .compose(order -> protectionService.isOperationRestricted(order.getAcqUnitIds(), DELETE, requestContext)));
   }
 
-
+  private Future<Void> createShadowInstanceIfNeeded(CompositePoLine compositePoLine, RequestContext requestContext) {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/helper/PurchaseOrderLineHelper.java:889]: Ensure that extra costs are not incurred for non-ECS flow. We should check if ECS is enabled before attempting to create shadow instances. This should exclude needing to make a call to mod-inventory to see if the shadow instance exists.

File: src/main/java/org/folio/service/orders/lines/update/OrderLinePatchOperationService.java [hunk 3]
```diff
@@ -63,17 +64,21 @@ public class OrderLinePatchOperationService {
...
     this.orderLinePatchOperationHandlerResolver = orderLinePatchOperationHandlerResolver;
     this.purchaseOrderLineService = purchaseOrderLineService;
     this.inventoryCache = inventoryCache;
+    this.inventoryManager = inventoryManager;
   }
 
   public Future<Void> patch(String lineId, PatchOrderLineRequest request, RequestContext requestContext) {
-    return patchOrderLine(request, lineId, requestContext)
+    String newInstanceId = request.getReplaceInstanceRef().getNewInstanceId();
+    return inventoryManager.createShadowInstanceIfNeeded(newInstanceId, requestContext)
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/service/orders/lines/update/OrderLinePatchOperationService.java:80]: check if ECS is enabled before performing shadow instance operations.

File: src/main/java/org/folio/service/consortium/ConsortiumConfigurationService.java (hunk 1 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/service/consortium/ConsortiumConfigurationService.java:20]: Thanks. Added

File: src/main/java/org/folio/service/consortium/SharingInstanceService.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.folio.rest.core.exceptions.ConsortiumException;
+import org.folio.rest.core.models.RequestContext;
+import org.folio.rest.core.models.RequestEntry;
+import org.folio.rest.tools.utils.TenantTool;
+
+import java.util.Map;
+import java.util.UUID;
+
+
+public class SharingInstanceService {
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply src/main/java/org/folio/service/consortium/SharingInstanceService.java:28]: Done

File: src/main/java/org/folio/helper/PurchaseOrderLineHelper.java (hunk 2 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/helper/PurchaseOrderLineHelper.java:889]: Done

File: src/main/java/org/folio/service/orders/lines/update/OrderLinePatchOperationService.java (hunk 3 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/service/orders/lines/update/OrderLinePatchOperationService.java:80]: Done

File: src/main/java/org/folio/service/consortium/SharingInstanceService.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  private Future<SharingInstance> shareInstance(String consortiumId, SharingInstance sharingInstance, RequestContext requestContext) {
+    RequestEntry requestEntry = new RequestEntry(SHARE_INSTANCE_ENDPOINT).withId(consortiumId);
+    return restClient.post(requestEntry, sharingInstance, SharingInstance.class, requestContext)
+      .compose(response -> {
+        if (ObjectUtils.notEqual(SharingStatus.ERROR, response.status())) {
+          logger.debug("Successfully sharedInstance with id: {}, sharedInstance: {}", response.instanceIdentifier(), response);
+          return Future.succeededFuture(response);
+        } else {
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/service/consortium/SharingInstanceService.java:63]: I think it makes sense to add 'error' in response or in logs fetched from SharingInstance entity.

File: src/main/java/org/folio/service/consortium/SharingInstanceService.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+          String message = String.format(SHARING_INSTANCE_ERROR, sharingInstance.sourceTenantId(), sharingInstance.targetTenantId(),
+            sharingInstance.instanceIdentifier());
+          return Future.failedFuture(new ConsortiumException(message));
+        }
+      });
+  }
+
+  private RequestContext createRequestContextWithUpdatedTenantId(Context context, Map<String, String> headers, String tenantId) {
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/service/consortium/SharingInstanceService.java:71]: maybe makes sence to rename tenantId to centralTenantId to give more context for person who looking into the code

File: src/main/java/org/folio/service/inventory/InventoryManager.java [hunk 7]
```diff
@@ -1265,4 +1277,24 @@ private void updateItemWithPieceFields(Piece piece, JsonObject item) {
...
+            .compose(instance -> {
+              if (Objects.nonNull(instance) && !instance.isEmpty()) {
+                return Future.succeededFuture();
+              }
+              logger.info("Creating shadow instance with instanceId: {}", instanceId);
+              return sharingInstanceService.createShadowInstance(instanceId, consortiumConfiguration.get(), requestContext);
+            });
+        }
+        logger.warn("ECS mode is not enabled for tenant with id {} when creating shadow instance with id {}",
+          TenantTool.tenantId(requestContext.getHeaders()), instanceId);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/service/inventory/InventoryManager.java:1295]: I think we don't need this warn log, if we are not in ECS mode - so consortiumConfiguration will be never presented and it is normal situation

File: src/main/java/org/folio/service/consortium/SharingInstanceService.java (hunk 5 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/service/consortium/SharingInstanceService.java:63]: Thanks for review, Added

File: src/main/java/org/folio/service/consortium/SharingInstanceService.java (hunk 6 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/service/consortium/SharingInstanceService.java]: Renamed

File: src/main/java/org/folio/service/inventory/InventoryManager.java (hunk 7 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/service/inventory/InventoryManager.java]: removed

File: pom.xml [hunk 8]
```diff
@@ -170,11 +170,6 @@
       <version>1.4.2</version>
       <type>pom</type>
     </dependency>
-    <dependency>
```
Comment (okolawole-ebsco) [inline pom.xml:173]: Why was this removed?

**Comment by Abdulkhakimov** [comment]
image

**Review by okolawole-ebsco — APPROVED** [review]

File: pom.xml (hunk 8 above)
Comment (Abdulkhakimov) [reply pom.xml:173]: @SerhiiNosko please explain. thank you

**Review by SerhiiNosko — APPROVED** [review]

File: pom.xml (hunk 8 above)
Comment (SerhiiNosko) [reply pom.xml:173]: Data processing core had folio-kafka-wrapper with 3.0.0-SNAPSHOT version and folio-kafka-wrapper had 2.6.0 version that causes conflicts. Now folio-kafka-wrapper is getting as transitive dependency from data-import-processing-core with 3.0.0-SNAPHOT version
